### PR TITLE
Bazel Build alterations

### PR DESCRIPTION
**Pull request by jsuereth**
This PR attempts to do the following:

- Abide by nested-workspace conventions in BAZEL community (see https://github.com/bazelbuild/bazel/issues/1943)
  _Note: This follows OpenCensusCpp implementation pretty closely_
- Update BAZEL to *also* use git submodules, keeping the build as close to CMAKE as possible.
  _Note: These submodules are NOT locked to version/tags, but really should be.  Also, lack of SHA verification has me concerned.  I'm not sure what the git submodule equivalent of this is._
- Update documentation to make sure developers know to init the submodules to build.

### CODE REVIEW

File: bazel/repository.bzl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        path = "third_party/benchmark",
+    )
+
+    # GoogleTest framework.
+    # Only needed for tests, not to build the OpenTelemetry library.
+    maybe(
+        native.local_repository,
+        name = "com_google_googletest",
+        path = "third_party/googletest",
+    )
```
Comment (lalitb) [inline bazel/repository.bzl:21]: Nice. We are using google-test and benchmark from submodules directory  now : )

**Review by lalitb — APPROVED** [review]
LGTM. Would like some more eye to review too due to my limited bazel knowledge : )

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -73,6 +73,14 @@ Add your fork as a remote:
...
 ```
 
+If you haven't, make sure you are loading the submodules required to build OpenTelemetry
+
+```sh
+git submodule init
+git submodule update
+```
+
+
```
Comment (ThomsonTan) [inline CONTRIBUTING.md:83]: nit: remove 1 new line as 1 new line is fine?

File: bazel/repository.bzl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        sha256 = "b679cef31102ed8beddc39ecfd6368ee311cbee6f50742f13f21be7278781821",
+        strip_prefix = "protobuf-3.11.2",
+        urls = [
+            "https://github.com/protocolbuffers/protobuf/releases/download/v3.11.2/protobuf-all-3.11.2.tar.gz",
+        ],
+    )
+
+    # OTLP Protocol definition
+    maybe(
+        native.new_local_repository,
```
Comment (ThomsonTan) [inline bazel/repository.bzl:48]: This seems coming from the `WORKSPACE` file, but still wondering why OTLP protocol definition is `native.new_local_repository` but others are `native.local_repository`?

Comment (jsuereth) [reply bazel/repository.bzl:48]: That's the difference between something that HAS a bazel-build already, or one where we're adapting it to be a BAZEL workspace.

OTLP does NOT have a bazel build, so we're provding our own build file.

File: WORKSPACE [hunk 4]
```diff
@@ -14,20 +14,13 @@
 
 workspace(name = "io_opentelemetry_cpp")
 
-load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive", "http_file")
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply WORKSPACE:17]: actually, it can be removed now that all repo loading is down in `.bzl` files.

**Review by ThomsonTan — APPROVED** [review]

File: bazel/repository.bzl [hunk 5]
```diff
@@ -0,0 +1,80 @@
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+def opentelemetry_cpp_deps():
+    """Loads dependencies need to compile the opentelemetry-cpp library."""
+
+    # Google Benchmark library.
+    # Only needed for benchmarks, not to build the OpenTelemetry library.
+    maybe(
+        native.local_repository,
```
Comment (g-easy) [inline bazel/repository.bzl:10]: Are consumers of otel-cpp meant to load e.g. benchmark before calling `opentelemetry_cpp_deps()`?

**Review by g-easy — APPROVED** [review]

File: bazel/repository.bzl (hunk 5 above)
Comment (jsuereth) [reply bazel/repository.bzl:10]: yeah.  I need to write the "INSTALLING" section for bazel, but basic gist is if you want to configure your own version/build that you can do so if you load BEFORE calling our macro, but you'll be on your own ensuring the version is compatible with us.

**Pull request merged**